File: Python/benchmark.py
import time
import random
import math
from typing import List, Dict, Tuple, Callable
from RSPCVT import tranthong, RSPCVT

def create_random_map(size: int, density: float) -> List[List[bool]]:
    """Generates a 2D grid where True represents a wall."""
    return [[random.random() < density for _ in range(size)] for _ in range(size)]

def run_benchmark():
    ITERATIONS = 10000  # Python is slower, so we use fewer iterations than LuaJIT
    RADIUS = 20
    MAP_SIZE = 60
    DENSITY = 0.25
    SEED = 42

    print(f"Benchmarking Python RSPCVT: Radius {RADIUS}, {ITERATIONS} iterations")
    print("-" * 50)

    # Pre-compute the Trie
    fov_engine = RSPCVT(RADIUS)
    
    # Pre-calculate the list of all relative targets for Naive
    targets = []
    for i in range(-RADIUS, RADIUS + 1):
        for j in range(-RADIUS, RADIUS + 1):
            if i*i + j*j <= RADIUS*RADIUS:
                targets.append((i, j))

    # 1. Baseline: Map Generation Overhead
    random.seed(SEED)
    start_map = time.perf_counter()
    for _ in range(ITERATIONS):
        create_random_map(MAP_SIZE, DENSITY)
    map_gen_total = time.perf_counter() - start_map
    print(f"Avg Map Gen Time:    {map_gen_total / ITERATIONS:.6f} s")

    # 2. Naive Dense Raycasting
    random.seed(SEED)
    start_naive = time.perf_counter()
    for _ in range(ITERATIONS):
        world_map = create_random_map(MAP_SIZE, DENSITY)
        ox, oy = 30, 30
        
        # Every single target tile gets a dedicated ray
        for tx_off, ty_off in targets:
            tx, ty = ox + tx_off, oy + ty_off
            is_obstructed = False
            
            def check_step(x, y):
                nonlocal is_obstructed
                if is_obstructed: return
                if 0 <= x < MAP_SIZE and 0 <= y < MAP_SIZE:
                    if world_map[y][x]:
                        is_obstructed = True

            tranthong(ox, oy, tx, ty, check_step)
    
    naive_total = (time.perf_counter() - start_naive) - map_gen_total

    # 3. RSPCVT FOV
    random.seed(SEED)
    start_trie = time.perf_counter()
    for _ in range(ITERATIONS):
        world_map = create_random_map(MAP_SIZE, DENSITY)
        ox, oy = 30, 30
        
        def blocks_vision(x, y):
            if 0 <= x < MAP_SIZE and 0 <= y < MAP_SIZE:
                return world_map[y][x]
            return True

        def set_visible(x, y):
            pass

        fov_engine.fov(ox, oy, blocks_vision, set_visible)

    trie_total = (time.perf_counter() - start_trie) - map_gen_total

    # Results
    print(f"Total Naive Time:    {naive_total:.4f} s (excl. map gen)")
    print(f"Total RSPCVT Time:   {trie_total:.4f} s (excl. map gen)")
    print("-" * 50)
    print(f"Speedup Factor:      {naive_total / trie_total:.2f}x faster")
    print(f"Avg RSPCVT FOV:      {trie_total / ITERATIONS:.6f} s")

if __name__ == "__main__":
    run_benchmark()

File: Lua/benchmark.lua
local RSPCVT = require 'RSPCVT'
local TranThong = require 'TranThong'

-- Use os.clock() for standard Lua timing, or socket.gettime() for precision
local getTime = os.clock

local ITERATIONS = 50000
local RADIUS = 20
local MAP_SIZE = 60
local DENSITY = 0.25

local fov_engine = RSPCVT(RADIUS)

-- Pre-calculate the list of all relative targets in the circle for the Naive approach
local targets = {}
for i = -RADIUS, RADIUS do
    for j = -RADIUS, RADIUS do
        if i*i + j*j <= RADIUS*RADIUS then
            table.insert(targets, {x = i, y = j})
        end
    end
end

local function create_random_map(size, density)
    local map = {}
    for y = 0, size do
        map[y] = {}
        for x = 0, size do
            map[y][x] = math.random() < density
        end
    end
    return map
end

local function run_benchmark()
    print(string.format("Benchmarking: Radius %d, %d iterations", RADIUS, ITERATIONS))
    print("--------------------------------------------------")

    -- 1. Measure Map Generation Overhead
    math.randomseed(42)
    local start_map = getTime()
    for _ = 1, ITERATIONS do
        create_random_map(MAP_SIZE, DENSITY)
    end
    local map_gen_total = getTime() - start_map
    local map_gen_avg = map_gen_total / ITERATIONS
    print(string.format("Avg Map Gen Time:    %.6f s", map_gen_avg))

    -- 2. Measure Naive Dense Raycasting
    math.randomseed(42)
    local start_naive = getTime()
    for _ = 1, ITERATIONS do
        local map = create_random_map(MAP_SIZE, DENSITY)
        local ox, oy = 30, 30
        
        -- For every single target tile, shoot a dedicated ray
        for k = 1, #targets do
            local t = targets[k]
            local tx, ty = ox + t.x, oy + t.y
            local blocked = false
            TranThong(ox, oy, tx, ty, function(x, y)
                if blocked then return end
                if map[y] and map[y][x] then
                    blocked = true
                end
                -- set_visible would happen here
            end)
        end
    end
    local naive_total = (getTime() - start_naive) - map_gen_total
    
    -- 3. Measure RSPCVT
    math.randomseed(42)
    local start_trie = getTime()
    for _ = 1, ITERATIONS do
        local map = create_random_map(MAP_SIZE, DENSITY)
        local ox, oy = 30, 30
        
        fov_engine:FOV(ox, oy, 
            function(x, y) return map[y] and map[y][x] end,
            function(x, y) --[[ set_visible --]] end
        )
    end
    local trie_total = (getTime() - start_trie) - map_gen_total

    -- Results
    print(string.format("Total Naive Time:    %.4f s (excl. map gen)", naive_total))
    print(string.format("Total RSPCVT Time:   %.4f s (excl. map gen)", trie_total))
    print("--------------------------------------------------")
    print(string.format("Speedup Factor:      %.2fx faster", naive_total / trie_total))
    print(string.format("Avg RSPCVT FOV:      %.6f s", trie_total / ITERATIONS))
end

run_benchmark()
